services/certmanager/standard/service.go

```diff
@@ -0,0 +1,168 @@
+
+// New creates a new cert manager service.
+func New(ctx context.Context, params ...Parameter) (*Service, error) {
+	parameters, err := parseAndCheckParameters(params...)
+	if err != nil {
+		return nil, errors.Wrap(err, "problem with parameters")
+	}
+
+	// Set logging.
+	log = zerologger.With().Str("service", "certmanager").Str("impl", "standard").Logger()
+	if parameters.logLevel != log.GetLevel() {
+		log = log.Level(parameters.logLevel)
+	}
+
+	// Load the certificates immediately.
+	certPEMBlock, err := parameters.majordomo.Fetch(ctx, parameters.certPEMURI)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to obtain server certificate")
+	}
+	certKeyBlock, err := parameters.majordomo.Fetch(ctx, parameters.certKeyURI)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to obtain server key")
+	}
+
+	// Initialise the certificate pair.
+	serverCert, err := tls.X509KeyPair(certPEMBlock, certKeyBlock)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to load certificate pair")
+	}
+	if len(serverCert.Certificate) == 0 {
+		return nil, errors.New("certificate file does not contain a certificate")
+	}
+	cert := serverCert.Leaf
+	if cert.NotAfter.Before(time.Now()) {
+		log.Warn().Time("expiry", cert.NotAfter).Msg("Server certificate expired")
+	}
+
+	log.Info().Str("issued_to", cert.Subject.CommonName).Str("issued_by", cert.Issuer.CommonName).Time("valid_until", cert.NotAfter).Msg("Server certificate loaded")
+
+	out := &Service{
+		ctx:             ctx,
+		majordomo:       parameters.majordomo,
+		certPEMURI:      parameters.certPEMURI,
+		certKeyURI:      parameters.certKeyURI,
+		reloadThreshold: parameters.reloadThreshold,
+		reloadInterval:  parameters.reloadInterval,
+	}
+	out.currentCert.Store(&serverCert)
+	return out, nil
+}
+
+func (s *Service) TryReloadCertificate() {
+	if !s.currentCertMutext.TryLock() {
+		// Certificate is already being reloaded; do nothing.
+		return
+	}
+	defer s.currentCertMutext.Unlock()
+
+	s.lastReloadAttemptTime = time.Now()
+
+	ctx := s.ctx
+	if s.reloadInterval > 0 {
+		var cancel context.CancelFunc
+		// Give up on the reload if it takes longer than the reload interval.
+		ctx, cancel = context.WithDeadline(s.ctx, s.lastReloadAttemptTime.Add(s.reloadInterval))
+		defer cancel()
+	}
+
+	certPEMBlock, err := s.majordomo.Fetch(ctx, s.certPEMURI)
+	if err != nil {
+		log.Warn().Err(err).Msg("Failed to obtain server certificate during reload")
+		return
+	}
+	certKeyBlock, err := s.majordomo.Fetch(ctx, s.certKeyURI)
+	if err != nil {
+		log.Warn().Err(err).Msg("Failed to obtain server key during reload")
+		return
+	}
+
+	// Load the certificate pair.
+	serverCert, err := tls.X509KeyPair(certPEMBlock, certKeyBlock)
+	if err != nil {
+		log.Warn().Err(err).Msg("Failed to load certificate pair during reload")
+		return
+	}
+	if len(serverCert.Certificate) == 0 {
+		log.Warn().Msg("Certificate file does not contain a certificate")
+		return
+	}
+	cert := serverCert.Leaf
```

Comment:
This also panicked, again I had to switch to:


```suggestion
	cert, err := x509.ParseCertificate(serverCert.Certificate[0])
	if err != nil || cert == nil {
		log.Warn().Msg("Failed to parse certificate")
		return
	}
```